optaplanner-core/src/test/java/org/optaplanner/core/api/score/stream/JoinersTest.java

```diff
@@ -300,9 +308,163 @@ public void greaterThanOrEqualPenta() {
                 (AbstractPentaJoiner<BigInteger, BigInteger, BigInteger, BigInteger, BigDecimal>) Joiners
                         .greaterThanOrEqual(leftMapping, rightMapping);
         assertSoftly(softly -> {
-            softly.assertThat(joiner.matches(TEN, ZERO, ZERO, ZERO, BigDecimal.TEN)).isTrue();
-            softly.assertThat(joiner.matches(TEN, ZERO, ZERO, ZERO, BigDecimal.ONE)).isTrue();
-            softly.assertThat(joiner.matches(ZERO, ZERO, ZERO, ZERO, BigDecimal.ONE)).isFalse();
+            softly.assertThat(joiner.matches(BigInteger.TEN, BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO,
+                    BigDecimal.TEN)).isTrue();
+            softly.assertThat(joiner.matches(BigInteger.TEN, BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO,
+                    BigDecimal.ONE)).isTrue();
+            softly.assertThat(joiner.matches(BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO, BigInteger.ZERO,
+                    BigDecimal.ONE)).isFalse();
+        });
+    }
+
+    private static final class Interval<A, B> {
+        final A start;
+        final A end;
+        final Function<A, B> map;
+
+        public Interval(A start, A end, Function<A, B> map) {
+            this.start = start;
+            this.end = end;
+            this.map = map;
+        }
+
+        public B getStart() {
+            return map.apply(start);
+        }
+
+        public B getEnd() {
+            return map.apply(end);
+        }
+
+        @Override
+        public String toString() {
+            return "(" + start.toString() + ", " + end.toString() + ")";
+        }
+
+        public static <A, B, C> IntervalFactory<A, B, C> createFactory(Function<C, A> constructorMap, Function<A, B> map) {
+            return new IntervalFactory<A, B, C>(constructorMap, map);
+        }
+
+        private final static class IntervalFactory<A, B, C> {
+            final Function<C, A> constructorMap;
+            final Function<A, B> map;
+
+            public IntervalFactory(Function<C, A> constructorMap, Function<A, B> map) {
+                this.constructorMap = constructorMap;
+                this.map = map;
+            }
+
+            public Interval<A, B> of(C start, C end) {
+                return new Interval<A, B>(constructorMap.apply(start), constructorMap.apply(end), map);
+            }
+        }
+    }
+
+    private final Interval.IntervalFactory<BigInteger, Long, Long> bigIntInterval =
+            Interval.createFactory(BigInteger::valueOf, BigInteger::longValue);
+
+    private final Interval.IntervalFactory<BigDecimal, Long, Long> bigDecimalInterval =
+            Interval.createFactory(BigDecimal::valueOf, BigDecimal::longValue);
```

Comment:
Personally, I think this a bit over-engineered. You don't need the factory, just use the `Interval` constructor directly.

These are tests - we're not looking for memory efficiency or performance, we're looking for maintainable code that is easy to understand.

Reply:
(To clarify: I prefer `BigDecimalInterval` and `BigIntegerInterval` to the factory that effectively does the same thing, only with a level of indirection.)

Reply:
I personally think calling assert with `new Interval(...)` doesn't look pretty, which was the main reason why I created the factories.

Reply:
Therefore my second comment - if you create factory methods directly on `Interval`, the code will still be easier to understand. The current `IntervalFactory` approach creates indirection that is harder to see through.